### PR TITLE
H3 plugin

### PR DESCRIPTION
**Pull request by GordonSmith**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

File: plugins/h3/lib_h3.ecllib [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+EXPORT h3_degrees_t := REAL8;
+EXPORT h3_rads_t := REAL8;
+EXPORT h3_resolution_t := UNSIGNED4;
+EXPORT h3_latlng_t := SET OF h3_degrees_t;
+EXPORT h3_boundary_t := SET OF h3_degrees_t;
+
+EXPORT h3 := SERVICE : plugin('h3'), library('h3'), namespace('h3'), CPP, PURE
+
+  //  Indexing  ---
+  h3_index_t index(CONST h3_degrees_t lat, CONST h3_degrees_t lng, CONST h3_resolution_t resolution) : cpp,action,context,fold,entrypoint='index';
```
Comment (GordonSmith) [inline plugins/h3/lib_h3.ecllib]: Does "CONST" mean anything for numerics?

File: plugins/h3/h3.hpp [hunk 2]
```diff
@@ -0,0 +1,53 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2016 HPCC Systems®.
```
Comment (GordonSmith) [inline plugins/h3/h3.hpp]: Year

File: plugins/h3/lib_h3.ecllib [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  SET OF h3_degrees_t center(CONST h3_index_t idx) : cpp,action,context,fold,entrypoint='center';
+  SET OF h3_degrees_t boundary(CONST h3_index_t idx) : cpp,action,context,fold,entrypoint='boundary';
+
+  //  Intorspection  ---
```
Comment (GordonSmith) [inline plugins/h3/lib_h3.ecllib]: Typo

**Comment by richardkchapman** [comment]
@GordonSmith Please rebase for 7.2.0

File: plugins/h3/lib_h3.ecllib (hunk 1 above)
Comment (richardkchapman) [reply plugins/h3/lib_h3.ecllib]: I don't think so.

File: plugins/h3/lib_h3.ecllib [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+EXPORT h3_resolution_t := UNSIGNED4;
+EXPORT h3_latlng_t := SET OF h3_degrees_t;
+EXPORT h3_boundary_t := SET OF h3_degrees_t;
+
+EXPORT h3 := SERVICE : plugin('h3'), library('h3'), namespace('h3'), CPP, PURE
+
+  //  Indexing  ---
+  h3_index_t index(CONST h3_degrees_t lat, CONST h3_degrees_t lng, CONST h3_resolution_t resolution) : cpp,action,context,fold,entrypoint='index';
+  SET OF h3_degrees_t center(CONST h3_index_t idx) : cpp,action,context,fold,entrypoint='center';
+  SET OF h3_degrees_t boundary(CONST h3_index_t idx) : cpp,action,context,fold,entrypoint='boundary';
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply plugins/h3/lib_h3.ecllib]: How big are the returned sets?

File: plugins/h3/h3.hpp [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+extern "C++"
+{
+    namespace h3
+    {
+    //--------------------------SET----------------------------------------
+    //ECL_H3_API uint64_t ECL_H3_CALL latLngToIndex(ICodeContext *_ctx, float lat, float lng, uint32_t resolution);
+    //ECL_H3_API void ECL_H3_CALL indexToString(ICodeContext *_ctx, size32_t &returnLength, char *&returnValue, unsigned __int64 index);
+    //ECL_H3_API unsigned __int64 ECL_H3_CALL stringToIndex(ICodeContext *_ctx, const char *stringIndex);
+    //ECL_H3_API void ECL_H3_CALL func2(ICodeContext *_ctx, size32_t &returnLength, char *&returnValue, unsigned __int64 param1, const char *param2, size32_t param3ValueLength, const char *param3Value);
+    } // namespace h3
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply plugins/h3/h3.hpp]: No I don't think so

File: plugins/h3/lib_h3.ecllib (hunk 4 above)
[the remark this answers is not in the document]
Comment (GordonSmith) [reply plugins/h3/lib_h3.ecllib]: 6 or 7 pairs.

[the remark this answers is not in the document]
Comment (GordonSmith) [reply plugins/h3/lib_h3.ecllib]: Ideally a SET OF <PAIR> is what I would like?

File: plugins/h3/h3.cpp [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+
+void toSetOf(bool &__isAllResult, size32_t &__lenResult, void *&__result, const H3Index *buff, int maxBuffLen)
+{
+    int len = 0;
+    while (len <= maxBuffLen && buff[len] != 0)
+        ++len;
+
+    __isAllResult = false;
+    __lenResult = len;
+    __result = rtlMalloc(len * sizeof(H3Index));
```
Comment (GordonSmith) [inline plugins/h3/h3.cpp]: Memory Leak

File: plugins/h3/h3.cpp [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+{
+    return h3GetBaseCell(index);
+}
+
+ECL_H3_API void ECL_H3_CALL toString(ICodeContext *_ctx, size32_t &lenVarStr, char *&varStr, unsigned __int64 index)
+{
+    char buff[17];
+    h3ToString(index, buff, 17);
+    lenVarStr = strlen(buff);
+    varStr = (char *)rtlMalloc(lenVarStr + 1);
```
Comment (GordonSmith) [inline plugins/h3/h3.cpp]: Memory Leak

**Comment by GordonSmith** [comment]
@richardkchapman @ghalliday I think this is ready for proper review - @ghalliday there is a question over a 6 row dataset versus set.

File: plugins/h3/h3.cpp [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+//  Indexing Functions ---
+ECL_H3_API unsigned __int64 ECL_H3_CALL index(ICodeContext *_ctx, double lat, double lng, uint32_t resolution)
+{
+    GeoCoord location;
+    location.lat = degsToRads(lat);
+    location.lon = degsToRads(lng);
+    return ::geoToH3(&location, resolution);
+}
+
+ECL_H3_API void ECL_H3_CALL center(ICodeContext *_ctx, bool &__isAllResult, size32_t &__lenResult, void *&__result, unsigned __int64 index)
```
Comment (ghalliday) [inline plugins/h3/h3.cpp]: Should possibly return a (lat, long) record, but I'm not sure how well records being returned from external calls are supported.

File: plugins/h3/README.md [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+h3_index := h3.index(40.689167, -74.044444, 15);
+```
+
+API
+----
+
+### Type Declarations
+```c
+EXPORT h3_index_t := UNSIGNED8;
+EXPORT h3_stringIndex_t := VARSTRING;
```
Comment (ghalliday) [inline plugins/h3/README.md]: not actually used.

**Review by ghalliday — COMMENTED** [review]
@GordonSmith potentially looks very useful.
A few minor comments.  Could also do with some test cases.  Do we want to include a wrapper in the standard library?

File: esp/src/lws.config.js [hunk 10]
```diff
@@ -1,4 +1,4 @@
-const debugServerIP = "192.168.99.103";
+const debugServerIP = "192.168.3.22";
```
Comment (richardkchapman) [inline esp/src/lws.config.js]: Should this be checked in?

Comment (GordonSmith) [reply esp/src/lws.config.js]: Nope

**Comment by ghalliday** [comment]
Probably best place for some plugin tests is in testing/regress/ecl along with the other plugin tests that are run as part of the regression suite.
It looks like the cmake file needs to incdicate that c99 is required to compile the c files.
@GordonSmith

**Comment by AttilaVamos** [comment]
@AttilaVamos Those spray errors has been caused by my merged PR-https://github.com/hpcc-systems/HPCC-Platform/pull/12206 . I'm working on the fix. I am sorry about inconvenience caused.

**Comment by AttilaVamos** [comment]
@GordonSmith back to you.

**Pull request closed without merging**